### PR TITLE
Improve the plugin build script with rollup

### LINKED ISSUE #296: Make asset bundling work with watch mode
We likely want to move to using webpack to bundle plugins to solve this.

We don't want any optimizations to be run on the build output though, leave that to the app build instead. The webpack config would be something inbetween a production config and a dev config. Similar to the `build:quick` command that Spotify has internally.

### PR DESCRIPTION
**Pull request by nikek**
## Hey, I just made a Pull Request!

Adding rollup to handle the build lets us handle css and assets in a consistent way together with typescript compiling and potential future additions.

Fixes https://github.com/spotify/backstage/issues/296

### CODE REVIEW

File: packages/cli/templates/default-plugin/tsconfig.json [hunk 1]
```diff
@@ -1,4 +1,17 @@
...
+    "lib": ["es6", "dom", "es2016", "es2017"],
+    "sourceMap": true,
+    "allowJs": true,
+    "jsx": "react",
+    "declaration": true,
+    "moduleResolution": "node",
+    "allowSyntheticDefaultImports": true,
+    "esModuleInterop": true
+  },
+  "include": ["src", "images.d.ts"],
```
Comment (Rugvip) [inline packages/cli/templates/default-plugin/tsconfig.json]: is images.d.ts missing from the PR?

File: packages/cli/templates/default-plugin/tsconfig.json [hunk 2]
```diff
@@ -1,4 +1,17 @@
...
+  "exclude": ["node_modules", "build"]
```
Comment (Rugvip) [inline packages/cli/templates/default-plugin/tsconfig.json]: is exclude really needed when we use include?

File: packages/cli/src/index.ts [hunk 3]
```diff
@@ -116,4 +116,6 @@ process.on('unhandledRejection', rejection => {
 });
 
 main(process.argv);
-// main([process.argv[0], process.argv[1], '--version']);
+// process.chdir('../../plugins/welcome');
+// console.log(`DEBUG: ${process.cwd()}`);
+// main([process.argv[0], process.argv[1], 'plugin:build']);
```
Comment (freben) [inline packages/cli/src/index.ts]: y

File: packages/cli/package.json [hunk 4]
```diff
@@ -32,6 +32,8 @@
     "@types/ora": "^3.2.0",
     "@types/react-dev-utils": "^9.0.4",
     "@types/recursive-readdir": "^2.2.0",
+    "@types/rollup-plugin-peer-deps-external": "2.2.0",
+    "@types/rollup-plugin-postcss": "2.0.0",
```
Comment (freben) [inline packages/cli/package.json]: use `^`?

File: plugins/home-page/tsconfig.json [hunk 5]
```diff
@@ -1,4 +1,7 @@
 {
   "extends": "../../tsconfig.json",
-  "include": ["src"]
+  "include": ["src"],
+  "compilerOptions": {
+    "module": "es6"
+  }
```
Comment (freben) [inline plugins/home-page/tsconfig.json]: should be added to the template too

Comment (nikek) [reply plugins/home-page/tsconfig.json]: Maybe this is unnecessary now that we switched back to commonjs. I'll have a look

File: packages/cli/src/index.ts (hunk 3 above)
Comment (nikek) [reply packages/cli/src/index.ts]: ✅

File: packages/cli/package.json (hunk 4 above)
Comment (nikek) [reply packages/cli/package.json]: ✅

File: plugins/home-page/tsconfig.json (hunk 5 above)
Comment (nikek) [reply plugins/home-page/tsconfig.json]: Still necessary, and it is in the template already.

File: packages/cli/src/commands/plugin/build.ts [hunk 6]
```diff
@@ -14,43 +14,17 @@
...
+    input: rollupConfig.input,
+    plugins: rollupConfig.plugins,
+  };
+  const outputOptions = rollupConfig.output;
 
-  await copyStaticAssets();
-  await run('tsc', args);
+  const bundle = await rollup.rollup(inputOptions);
+  await bundle.generate(outputOptions);
+  await bundle.write(outputOptions);
```
Comment (Rugvip) [inline packages/cli/src/commands/plugin/build.ts]: This doesn't support watch mode right? We need support for that for ´yarn start´ to work

File: packages/cli/src/commands/plugin/rollup.config.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import peerDepsExternal from 'rollup-plugin-peer-deps-external';
+import typescript from 'rollup-plugin-typescript2'; // @rollup/plugin-typescript
```
Comment (Rugvip) [inline packages/cli/src/commands/plugin/rollup.config.ts]: why the comment?

File: plugins/home-page/tsconfig.json (hunk 5 above)
Comment (Rugvip) [reply plugins/home-page/tsconfig.json]: Can we move it to the root config?

File: packages/cli/src/commands/plugin/rollup.config.ts (hunk 7 above)
Comment (nikek) [reply packages/cli/src/commands/plugin/rollup.config.ts]: removed

**Review by Rugvip — APPROVED** [review]
Merge and iterate :grin:

Could you create issues for the known stuff, like export syntax not working correctly?

**Pull request merged**